api/src/main/java/edu/cornell/mannlib/vitro/webapp/application/VitroHomeDirectory.java

```diff
@@ -52,6 +73,219 @@ public String getDiscoveryMessage() {
 		return discoveryMessage;
 	}
 
+	/**
+	 * Populates VIVO home directory with files required to run.
+	 * 
+	 * NOTE: Will not overwrite any modified files on redeploy.
+	 */
+	public void populate() {
+		File vhdDir = getPath().toFile();
+
+		if (!vhdDir.isDirectory() || vhdDir.list() == null) {
+			throw new RuntimeException("Application home dir is not a directory! " + vhdDir);
+		}
+
+		Map<String, String> digest = untar(vhdDir);
+
+		writeDigest(digest);
+	}
+
+	/**
+	 * A non-destructive untar process that returns checksum digest of tarred files.
+	 * 
+	 * Checksum digest can be manually created with the following command.
+	 * 
+	 * `find /vivo/home -type f | grep -E "^/vivo/home/bin/|^/vivo/home/config/|^/vivo/home/rdf/" | xargs md5sum > /vivo/home/digest.md5`
+	 * 
+	 * @param destination VIVO home directory
+	 * @return digest of each files checksum
+	 */
+	private Map<String, String> untar(File destination) {
+		log.info("Syncing VIVO home at: " + destination.getPath());
+
+		Map<String, String> digest = new HashMap<>();
+		Map<String, String> storedDigest = loadDigest();
+
+		TarArchiveEntry tarEntry;
+		try (
+			InputStream homeDirTar = getHomeDirTar();
+			TarArchiveInputStream tarInput = new TarArchiveInputStream(homeDirTar);
+		) {
+			while ((tarEntry = tarInput.getNextTarEntry()) != null) {
+
+				// Use the example configurations
+				String outFilename = tarEntry.getName().replace("example.", "");
+				File outFile = new File(destination, outFilename);
+
+				// Is the entry a directory?
+				if (tarEntry.isDirectory()) {
+					if (!outFile.exists()) {
+						outFile.mkdirs();
+					}
+				} else {
+					// Entry is a File
+					boolean write = true;
+
+					// reading bytes into memory to avoid having to unreliably reset stream
+					byte[] bytes = IOUtils.toByteArray(tarInput);
+					String newFileChecksum = checksum(bytes);
+					digest.put(outFilename, newFileChecksum);
+
+					// if file already exists and stored digest contains the file,
+					// check to determine if it has changed
+					if (outFile.exists() && storedDigest.containsKey(outFilename)) {
+						String existingFileChecksum = checksum(outFile);
+						// if file has not changed in home and is not the same as new file, overwrite
+						write = storedDigest.get(outFilename).equals(existingFileChecksum)
```

Comment:
@wwelling if a file in custom VIVO_HOME has been changed, and checksum digest hasn't been manually created with the command ( `find /vivo/home -type f | grep -E "^/vivo/home/bin/|^/vivo/home/config/|^/vivo/home/rdf/" | xargs md5sum > /vivo/home/digest.md5`), the expression	 
`storedDigest.get(outFilename).equals(existingFileChecksum)`
will return false, meaning `write` will be false, and the file wouldn't be overwritten. It means it will work as described in the comment above  `if file **has not** changed in home and is not the same as new file, overwrite`. I can't find the code which will delete newly added files in the custom VIVO_HOME or overwrite customized files in custom VIVO_HOME during the VIVO upgrade. Can you please point me to that code, if exists? If it doesn't exist, why manual creation of digest is important?

Reply:
Files are never deleted. The untarring process is non-destructive. The purpose of the digest file is to track changes to files in the VIVO home directory. If a digest file does not exist when migrating from an existing instance and implementer intends to keep their changes to any VIVO home directory files, they should manually create the digest file.

Migration process of a custom instance will certainly have nuances all the same with customizations. The digest provides a mechanism to maintain customizations while informing in the log which files have been preserved from version to version. If no customizations, the defaults from the version migrating to will be upgraded.